components/post_view/post_attachment_opengraph/post_attachment_opengraph.tsx

```diff
@@ -41,230 +47,212 @@ export type Props = {
+    url?: string;
+    secure_url?: string;
+    type?: string;
+    width: number;
+    height: number;
+    format: string;
+    frameCount: number;
+}
 
-        const imageRatio = width / height;
+export function getBestImage(openGraphData?: OpenGraphMetadata, imagesMetadata?: Record<string, PostImage>) {
+    if (!openGraphData?.images?.length) {
+        return null;
+    }
 
-        return width >= largeImageMinWidth && imageRatio >= largeImageMinRatio;
-    };
+    // Get the dimensions from the post metadata if they weren't provided by the website as part of the OpenGraph data
+    const images = openGraphData.images.map((image: OpenGraphMetadataImage) => {
+        const imageUrl = image.secure_url || image.url;
 
-    renderImageToggle() {
-        if (this.props.isInPermalink) {
-            return undefined;
-        }
+        return {
+            ...image,
+            height: image.height || imagesMetadata?.[imageUrl].height || -1,
+            width: image.width || imagesMetadata?.[imageUrl].width || -1,
+            format: image.format || image.type?.split('/')[1] || image.type || '',
+            frameCount: 0,
+        };
+    });
 
-        return (
-            <button
-                className={'style--none post__embed-visibility color--link'}
-                data-expanded={this.props.isEmbedVisible}
-                aria-label='Toggle Embed Visibility'
-                onClick={this.props.toggleEmbedVisibility}
-            />
-        );
-    }
+    return getNearestPoint<ImageData>(DIMENSIONS_NEAREST_POINT_IMAGE, images);
+}
 
-    renderLargeImage(imageUrl: string, imageMetadata?: PostImage) {
-        if (!this.props.isEmbedVisible) {
-            return undefined;
-        }
+export const getIsLargeImage = (data: ImageData) => {
+    const {height, width} = data;
 
-        return (
-            <ExternalImage
-                src={imageUrl}
-                imageMetadata={imageMetadata}
-            >
-                {(safeImageUrl: string) => (
-                    <SizeAwareImage
-                        className='attachment__image attachment__image--opengraph large_image'
-                        src={safeImageUrl}
-                        dimensions={imageMetadata}
-                    />
-                )}
-            </ExternalImage>
-        );
-    }
+    return width >= LARGE_IMAGE_WIDTH && (width / height) >= LARGE_IMAGE_RATIO;
+};
 
-    renderSmallImage(imageUrl: string, imageMetadata?: PostImage) {
-        if (!this.props.isEmbedVisible) {
-            return undefined;
+const PostAttachmentOpenGraph = ({openGraphData, post, actions, link, ...rest}: Props) => {
+    const [imageUrl, setImageUrl] = useState<string>();
+    const [isLargeImage, setIsLargeImage] = useState<boolean>(false);
+    const [showImagePreview, setShowImagePreview] = useState<boolean>(true);
+    const [previewRemoved, setPreviewRemoved] = useState<boolean>(post?.props?.[PostTypes.REMOVE_LINK_PREVIEW] === 'true');
```

Comment:
Do `imageUrl`, `isLargeImage`, and `previewRemoved` need to be stored in state, or can they be derived from the props each time this is called?

Reply:
I simplified it a lot by moving most of the logic to the actual Image component